cpp/src/rolling/range_rolling.cu

```diff
@@ -0,0 +1,234 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/aggregation.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/groupby/sort_helper.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/rolling.hpp>
+#include <cudf/reduction.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/rolling/range_window_bounds.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cub/device/device_segmented_reduce.cuh>
+#include <thrust/functional.h>
+#include <thrust/reduce.h>
+
+#include <optional>
+
+namespace CUDF_EXPORT cudf {
+namespace detail {
+
+/**
+ * @brief Compute the number of nulls in each group.
+ *
+ * @param orderby Column with null mask.
+ * @param offsets Offset array defining the (sorted) groups.
+ * @param stream CUDA stream used for kernel launches
+ * @return device_uvector containing the null count per group.
+ */
+[[nodiscard]] rmm::device_uvector<cudf::size_type> nulls_per_group(
+  column_view const& orderby,
+  rmm::device_uvector<size_type> const& offsets,
+  rmm::cuda_stream_view stream)
+{
+  auto d_orderby        = column_device_view::create(orderby, stream);
+  auto const num_groups = offsets.size() - 1;
+  std::size_t bytes{0};
+  auto is_null_it = cudf::detail::make_counting_transform_iterator(
+    cudf::size_type{0}, [orderby = *d_orderby] __device__(size_type i) -> size_type {
+      return static_cast<size_type>(orderby.is_null_nocheck(i));
+    });
+  rmm::device_uvector<cudf::size_type> null_counts{num_groups, stream};
+  cub::DeviceSegmentedReduce::Sum(nullptr,
+                                  bytes,
+                                  is_null_it,
+                                  null_counts.begin(),
+                                  num_groups,
+                                  offsets.begin(),
+                                  offsets.begin() + 1,
+                                  stream.value());
+  auto tmp = rmm::device_buffer(bytes, stream);
+  cub::DeviceSegmentedReduce::Sum(tmp.data(),
+                                  bytes,
+                                  is_null_it,
+                                  null_counts.begin(),
+                                  num_groups,
+                                  offsets.begin(),
+                                  offsets.begin() + 1,
+                                  stream.value());
+  return null_counts;
+}
```

Comment:
This is the major performance improvement in this PR. It makes it much faster to find the nulls per group in a grouped rolling window with low-cardinality group key and nulls in the orderby column (in bad cases the old code would take more than a second, this new code takes a few ms).